### DISCUSSION TITLE
Discussion: Whether to use `auto` for iterators

### DISCUSSION THREAD

**Discussion by stephengtuggy** [vegastrike/Vega-Strike-Engine-Source#800] [General]
I'm fine with type inference where it is trivial. Example from other languages is something like

`let someString = String('blah blah')`

However, I'm less comfortable with replacing _std::list<CollidableBackref>::iterator_ with _auto_. I prefer to clearly see my C++ abominations.

_Originally posted by @royfalk in https://github.com/vegastrike/Vega-Strike-Engine-Source/pull/789#discussion_r1324999186_

**Comment by BenjamenMeyer** [comment]
In general, I highly prefer being specific with types - even return types - and really don't like `auto` as a type.
However, one of the few cases where I find it okay is with dealing with the iterator types of objects since:
```
std::vector<someObject>::iterator
```
can be quite bother some to type out regularly. So using `auto` in that case can be acceptable, but I wouldn't say that's necessarily true of every case it might be used.

For example, looping over an iterable should generally be done in a relatively short block so the type should be generally evident especially since you're iterating over it so `auto` would likely be used . The following is a good use case:
```
for (auto entry =  myIterableType.begin();  entry != myIterableType.end(); ++entry) {
    ...
    // do something
   ...
}
```
However, if you're randomly accessing the iterator where it's going to get handed around or manipulated outside of a loop then `auto` really isn't good and type enforcement should be more strict.
```
iterableType::iterator instanceObject = myIterableType.begin();
// do something outside of the loop.
...
```
Make sense?

**Reply by stephengtuggy** [reply]
Makes sense to me.
